### ISSUE TITLE
feat(mdCompiler): support `preAssignBindingsEnabled()`

### ISSUE THREAD

**Issue by gkalpak** [angular/material#10016]
**Actual Behavior**:
 -  `What is the issue? *` `mdCompiler` [always pre-assigns bindings][1] on controller instances.
 -  `What is the expected behavior?` `mdCompiler` should respect the `preAssignBindingsEnabled()` flag (if present).
    This new configuration option was introduced with angular/angular.js#15095. Although it is not directly related to `mdCompiler`, people might expect the option to be respected by `ngMaterial` as well (possibly as a result of not realizing `ngMaterial` uses its own compiler for certain things).

**Angular Versions**: *
 -  `Angular Version:` 1.5.9+, 1.6.+
 -  `Angular Material Version:` All

[1]: https://github.com/angular/material/blob/f6dedff67dac455a91ac0a957b902f9274492e6a/src/core/services/compiler/compiler.js#L253-L255

**Comment by gkalpak** [comment]
For reference, here is how we fixed the same issue for `ngMock/$controller`: [angular@7d9a791][1]

--
And while you are at it, you might want to fix support for classes and controllers that return a value from their constructor ([demo of the issue][2]), by re-assigning the bindings if `ctrl !== invokeCtrl.instance` [here][3].

[1]: https://github.com/angular/angular.js/commit/7d9a791c6a8c80d29d6c84afa287c81f2a307439?w=1#diff-2a255ed5e9564e25ce6eb711b604f40fR2249
[2]: https://codepen.io/ExpertSystem/pen/bBwYxx?editors=1010
[3]: https://github.com/angular/material/blob/f6dedff67dac455a91ac0a957b902f9274492e6a/src/core/services/compiler/compiler.js#L258

**Comment by mgol** [comment]
Doing this right would also mean implementing support for this flag (& lifecycle hooks) to dialog controllers. But that, in turn, would be better achieved by dialogs just using real components instead of manually passing a controller and a template.

**Comment by gkalpak** [comment]
I agree that it is much better to leverage the core `$compile` service as much as possible (for feature parity, consistency and DRYness) and using components for dialogs (via templates) is an indirect way to achieve that (so totally recommended if you ask me).

In the meantime though, taking the value of `preAssignBindingsEnabled` into account requires very little effort, so I don't see why not.

(Note that although support for lifecycle hooks would be nice-to-have, it is not strictly necessary for this. AThat said, the `$onInit()` would be very helpful indeed.)

**Comment by mgol** [comment]
To take `preAssignBindingsEnabled` into account you need to at least implement the `$onInit` hook. Otherwise there's no place in the controller where referencing the bindings is safe.

**Comment by gkalpak** [comment]
It is safe referencing them in templates and in methods registered as event handlers or invoked via the template. I admit that in order to unlock the full potential you need the `$onInit` hook, but you can for example support some usecases, such as setting default values in the controller constructor and letting them be overwritten by bindings.

**Comment by mgol** [comment]
@gkalpak The issue I have with this partial approach is that I'd like to have `preAssignBindingsEnabled` set to `false` in Angular but I still need to apply initialization logic to my bindings in dialog controllers. I can work with any of the following scenarios:

1. Material doesn't respect the flag
2. Material respects the flag & implements `$onInit`

In the second, I'll write the logic in `$onInit` as in the regular components. In the first, I'll just remember I have to write it in a constructor in dialogs.

If Material respected the flag but didn't implement `$onInit`, I'm blocked. Hence, as long as there's no way to flip the flag only for some components, I think implementing taking this flag into account should be blocked on implementing `$onInit`.

**Comment by mgol** [comment]
BTW, if this issue is about Material following the flag as set in Angular then I think it requires a breaking change so can't arrive before 1.2.0. Unless the behavior was itself hidden behind a Material flag.

**Comment by gkalpak** [comment]
> Hence, as long as there's no way to flip the flag only for some components, I think implementing taking this flag into account should be blocked on implementing `$onInit`.

Fair point.

> BTW, if this issue is about Material following the flag as set in Angular then I think it requires a breaking change.

I am not sure I agree. The way I think about it, the fact that `ngMaterial` is manually instantiating the Dialog controllers and wiring everything together is an implementation detail. From the developer's point of view, the dialog configuration looks like a regular DDO on steroids (because of the extra features). The fact that `ngMaterial` tries to replicate the behavior of regular directives (e.g. by pre-assigning bindings on the controller instance), for example, reinforces this impression.

In that sense, I would consider it a breaking change in Angular not in `ngMaterial`. Although admittedly, this is one of those foggy, undocumented areas where things aren't black and white :smiley:

**Comment by mgol** [comment]
If Angular Material changes its behavior, breaking some apps that only updated Material, it can hardly be an Angular breaking change.

And this will break apps. So unless this feature lands before Angular 1.6 final and 1.5.10, i.e. versions that implement the `preAssignBindingsEnabled` flag (which is not realistic), this will be a breaking change in Angular Material, and as such can't land in a patch release but has to wait for 1.2.0.

**Comment by gkalpak** [comment]
> If Angular Material changes its behavior, breaking some apps that only updated Material and not Angular itself, it can hardly be an Angular breaking change.

Why would the behavior of `ngMaterial` change in that scenario? If the Angular version does not have the `preAssignBindingsEnabled` flag, then pre-assigning will be enabled. Basically, `ngMaterial` would have a condition like:

```js
if (angular.isFunction($compileProvider.preAssignBindingsEnabled) && !$compileProvider.preAssignBindingsEnabled()) {
  // No pre-assignment
} else {
  // Yes pre-assignment
}
```

**Comment by mgol** [comment]
As I said, the problematic behavior is when Angular 1.6.0 comes out before
this issue is fixed. One upgrades Angular to 1.6.0, adjusting the code
where needed and uses it with Angular Material 1.1.2. Then 1.1.3 with
support for the flag arrives, the developer automatically upgrades to it
and suddenly their app is broken. Sounds like a breaking change to me...

**Comment by gkalpak** [comment]
Sigh... Libraries and their backwards compatibility 😞
I agree adding this to a breaking release is the best option.

**Comment by mgol** [comment]
@ThomasBurleson It seems we've come to the conclusion this should have the "Breaking Change" label. Also, it doesn't critical to me so the `P0` flag may be too much?

**Comment by ThomasBurleson** [comment]
@here - shout-out to @DevVersion for an interesting PR.

**Comment by ThomasBurleson** [comment]
Refs PR #10469
